pretyping/nativenorm.ml

```diff
@@ -26,6 +26,68 @@ module RelDecl = Context.Rel.Declaration
 
 exception Find_at of int
 
+(* profiling *)
+
+let profiling_enabled = ref false
+    
+(* for supported platforms, filename for profiler results *)
+
+let profile_filename = ref "native_compute_profile.data"
+
+let profiler_supported_platforms = ["Unix (Linux)"]
+
+let profiler_platform () =
+  match [@warning "-8"] Sys.os_type with
+  | "Unix" ->
+     let in_ch = Unix.open_process_in "uname" in
+     let uname = input_line in_ch in
+     let _ = close_in in_ch in
+     Format.sprintf "Unix (%s)" uname
+  | "Win32" -> "Windows (Win32)"
+  | "Cygwin" -> "Windows (Cygwin)"
+
+let profiling_supported () = List.mem (profiler_platform ()) profiler_supported_platforms
+    
+let get_profile_filename () = !profile_filename
+
+let set_profile_filename fn =
+  if profiling_supported () then
+    profile_filename := fn
+
+(* counter so we don't overwrite existing profiles *)
+let profile_filename_counter = ref 0
+(* flag to guard against (unlikely) counter wraparound *)
+let profile_saw_zero_counter = ref false
+  
+(* find unused profile filename *)
+let get_available_profile_filename () =
+  let basename = get_profile_filename () in
+  let rec find_name () =
+    let ct = !profile_filename_counter in
+    let _ = assert (not (ct = 0 && !profile_saw_zero_counter)) in
+    let possible_name =
+      if ct = 0 then (
+	profile_saw_zero_counter := true;
+	basename )
+      else basename ^ "." ^ (string_of_int ct) in
+    if Sys.file_exists possible_name
+    then (
+      incr profile_filename_counter;
+      find_name ()
+    )
+    else possible_name
+  in
+  let name = find_name () in
+  incr profile_filename_counter;
+  name
+	
+let get_profiling_enabled () =
+  !profiling_enabled
+    
+let set_profiling_enabled b =
+  if profiling_supported () || not b then
+    profiling_enabled := b
+    
```

Comment:
You may want to optimize this a bit more, in particular no point on doing the expensive profiling supported so many times, and also if the profiling is already enabled/disabled.